judo/utils/normalization.py

```diff
@@ -0,0 +1,208 @@
+        """Initialize the normalizer.
+
+        Args:
+            dim: Dimension of the data.
+        """
+        self.dim = dim
+
+    def normalize(self, x: np.ndarray) -> np.ndarray:
+        """Normalize the data.
+
+        Args:
+            x: The data to normalize. Shape=(..., dim).
+        """
+        raise NotImplementedError
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Denormalize the data.
+
+        Args:
+            x: The data to denormalize. Shape=(..., dim).
+        """
+        raise NotImplementedError
+
+    def update(self, x: np.ndarray) -> None:
+        """Update the normalizer.
+
+        Args:
+            x: The data to update the normalizer with. Shape=(batch_size, dim).
+        """
+        pass
+
+
+class IdentityNormalizer(Normalizer):
+    """Normalizer that does nothing."""
+
+    def __init__(self, dim: int) -> None:
+        """Initialize the normalizer."""
+        super().__init__(dim)
+
+    def normalize(self, x: np.ndarray) -> np.ndarray:
+        """Return the data as is."""
+        return x
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Return the data as is."""
+        return x
+
+
+class MinMaxNormalizer(Normalizer):
+    """Normalizer that uses min and max values to scale the data to the range [-1, 1]."""
+
+    def __init__(self, dim: int, min: np.ndarray, max: np.ndarray, eps: float = 1e-8) -> None:
+        """Initialize the normalizer.
+
+        Args:
+            dim: Dimension of the data.
+            min: The minimum values of the data.
+            max: The maximum values of the data.
+            eps: Small value to prevent division by zero.
+        """
+        super().__init__(dim)
+        self.min = min
+        self.max = max
+        self.eps = eps
+
+        # Only normalize the dimensions that are not -inf or inf
+        self.norm_dims = np.where((self.min != -np.inf) & (self.max != np.inf))[0]
+
+    def normalize(self, x: np.ndarray) -> np.ndarray:
+        """Normalize the data."""
+        x_normalized = x.copy()
+        x_normalized[..., self.norm_dims] = (
+            2
+            * (x[..., self.norm_dims] - self.min[self.norm_dims])
+            / (self.max[self.norm_dims] - self.min[self.norm_dims])
+            - 1
+        )
+        return x_normalized
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Denormalize the data."""
+        x_denormalized = x.copy()
+        x_denormalized[..., self.norm_dims] = (x[..., self.norm_dims] + 1) * (
+            self.max[self.norm_dims] - self.min[self.norm_dims]
+        ) / 2 + self.min[self.norm_dims]
+        return x_denormalized
+
+
+class RunningMeanStdNormalizer(Normalizer):
+    """Normalizer that uses running statistics (mean and std) to scale the data."""
+
+    def __init__(
+        self,
+        dim: int,
+        init_std: float = 1.0,
+        min_std: float = 1e-6,
+        max_std: float = 1e6,
+        eps: float = 1e-8,
```

Comment:
i think eps should be a bit larger. we're using a default 64 bit precision for the numpy ops right now, but in the case where we allow torch ops and stuff later on, this could be a source of bugs, because it would eval to 0 at 32 bit precision (i think).